### PR TITLE
Feat: Implement middleware to check login sessions and redirect not logged in user

### PR DESCRIPTION
**Pull request by kotaaaa**
## Overview

Fix for #143

## Changes

- Added Redirect to top page if user is not login with Next.js middleware
- Added root path redirection 
  - Login user: "/" redirects to "/groups"
  - Non-login user: "/" redirects to "/login"
- If a logged in user accesses "/login", redirect to "/groups"

## Review points
- Please check Reviewer Checklist below

## Notes 
- when you want to develop after this PR merge, you have to login any google account or comment out `middleware.ts` to see food, groups, profile pages.

## Reference
   - [Amplify Dev Center Doc (for authentiation at middleware.ts)](https://docs.amplify.aws/react/build-a-backend/server-side-rendering/nextjs/#authentication-with-nextjs-server-side-runtime)
   - [Building fast Next.js apps using TypeScript and AWS Amplify JavaScript v6](https://aws.amazon.com/jp/blogs/mobile/amplify-javascript-v6/)

## Assignee Checklist:

- [x] The base branch is correct (See: [Types of Branches](https://github.com/genesis-tech-tribe/nishiki-frontend/blob/develop/docs/CONTRIBUTING.md#types-of-branches))
- [x] The branch name follows the [Branch Naming Conventions](https://github.com/genesis-tech-tribe/nishiki-frontend/blob/develop/docs/CONTRIBUTING.md#branch-naming-conventions)
- [x] The correct assignees and reviewers have been designated for this PR
- [x] The coding style follows the [Coding Style Guide](https://github.com/genesis-tech-tribe/nishiki-frontend/blob/develop/docs/STYLEGUIDE.md)
- [x] All the related issues are associated with this PR
- [x] All criteria in the associated issues are met (please tick the checkboxes)
- [x] My changes do not generate new warnings or errors (especially in the console)

## Reviewer Checklist:

- [x] The code follows the generic best practices and our [Coding Style Guide](https://github.com/genesis-tech-tribe/nishiki-frontend/blob/develop/docs/STYLEGUIDE.md)
- [x] Code is well-commented and easy to understand
- [x] UI changes accurately reflect the provided design

### CODE REVIEW

File: src/utils/amplifyServerUtils.ts [hunk 1]
```diff
@@ -0,0 +1,27 @@
+import { createServerRunner } from '@aws-amplify/adapter-nextjs';
+const OAUTH_DOMAIN: string = process.env.NEXT_PUBLIC_OAUTH_DOMAIN || '';
+const LOCALHOST_URL: string = process.env.NEXT_PUBLIC_LOCALHOST_URL || '';
+const OAUTH_REDIRECT_URL: string = process.env.NEXT_PUBLIC_OAUTH_REDIRECT_URL || '';
+const USER_POOL_ID: string = process.env.NEXT_PUBLIC_USER_POOL_ID || '';
+const USER_POOL_CLIENT_ID: string = process.env.NEXT_PUBLIC_USER_POOL_CLIENT_ID || '';
+
+// eslint-disable-next-line @typescript-eslint/no-unsafe-call
```
Comment (kotaaaa) [inline src/utils/amplifyServerUtils.ts:8]: https://github.com/aws-amplify/amplify-js/blob/04627618afe526dd776ee678b723ed6e697f440d/packages/adapter-nextjs/src/types/NextServer.ts#L89-L91

This method's type seems not to be open for external so I cannot help setting this ignore of eslint check for now.

https://docs.amplify.aws/nextjs/build-a-backend/server-side-rendering/nextjs-v5-to-v6-migration-guide/#using-the-api-category-in-v6

File: src/middleware.ts [hunk 2]
```diff
@@ -0,0 +1,41 @@
+import { runWithAmplifyServerContext } from '@/utils/amplifyServerUtils';
+
+import { fetchAuthSession } from 'aws-amplify/auth/server';
+import { NextRequest, NextResponse } from 'next/server';
+
+export async function middleware(request: NextRequest) {
+  const response = NextResponse.next();
+  // eslint-disable-next-line @typescript-eslint/no-unsafe-call
```
Comment (kotaaaa) [inline src/middleware.ts:8]: [This method's](https://github.com/aws-amplify/amplify-js/blob/04627618afe526dd776ee678b723ed6e697f440d/packages/adapter-nextjs/src/types/NextServer.ts#L86) type seems not to be open for external so I cannot help setting this ignore of eslint check for now.
[Example](https://docs.amplify.aws/nextjs/build-a-backend/server-side-rendering/nextjs-v5-to-v6-migration-guide/#step-2-server-side-configuration)

File: src/components/ui/LoginForm.tsx [hunk 3]
```diff
@@ -72,18 +51,24 @@ export const LoginForm = () => {
 
   async function currentSession() {
     try {
-      const { accessToken, idToken } = (await fetchAuthSession()).tokens ?? {};
-      console.log('accessToken, idToken', accessToken, idToken);
+      const { idToken } = (await fetchAuthSession()).tokens ?? {};
+      console.log('idToken', idToken?.toString());
```
Comment (kotaaaa) [inline src/components/ui/LoginForm.tsx:52]: This is how we get idToken for API gateway authentication. I will delete it after complete common API client function.

**Review by kanta1207 — COMMENTED** [review]
LGTM! Let's wait a bit for Yusuke's review!

File: src/middleware.ts
Comment (kotaaaa) [inline src/middleware.ts:1]: Reference: https://docs.amplify.aws/gen2/build-a-backend/server-side-rendering/#authentication-with-nextjs-server-side-runtime

File: src/components/ui/LoginForm.tsx [hunk 4]
```diff
@@ -1,89 +1,64 @@
...
-});
+  { ssr: true },
+);
 
 export const LoginForm = () => {
-  const [user, setUser] = useState<AuthUser | null>(null);
-  const [, setError] = useState<unknown>(null);
-  const [, setCustomState] = useState<string | null>(null);
+  const router = useRouter();
+  const [isLoading, setLoading] = useState(true);
```
Comment (kotaaaa) [inline src/components/ui/LoginForm.tsx:37]: Reference: https://docs.amplify.aws/react/build-a-backend/auth/add-social-provider/#set-up-your-frontend

File: src/components/ui/LoginForm.tsx [hunk 5]
```diff
@@ -1,89 +1,64 @@
...
         },
+        userPoolId: USER_POOL_ID,
+        userPoolClientId: USER_POOL_CLIENT_ID,
       },
-      userPoolId: USER_POOL_ID,
-      userPoolClientId: USER_POOL_CLIENT_ID,
     },
   },
-});
+  { ssr: true },
```
Comment (kotaaaa) [inline src/components/ui/LoginForm.tsx:31]: Reference: https://docs.amplify.aws/react/build-a-backend/server-side-rendering/nextjs/#configure-amplify-library-for-client-side-usage

> NOTE: To use the Amplify library on the client side in a Next.js app, you will need to set ssr to true when calling Amplify.configure. This instructs the Amplify library to store tokens in the cookie store of a browser. Cookies will be sent along with requests to your Next.js server for authentication.

**Review by uskayyyyy — APPROVED** [review]
@kotaaaa 
LGTM :) Thank you for your work!!

**Pull request merged**